AncestralAllele.pm

```diff
@@ -0,0 +1,277 @@
+
+ Note the first time you run the plugin with a newly generated FASTA file it will
+ spend some time indexing the file. DO NOT INTERRUPT THIS PROCESS, particularly
+ if you do not have Bio::DB::HTS installed.
+ 
+=cut
+
+package AncestralAllele;
+
+use strict;
+use warnings;
+
+use Bio::EnsEMBL::Utils::Sequence qw(reverse_comp);
+
+use Bio::EnsEMBL::Variation::Utils::BaseVepPlugin;
+
+use base qw(Bio::EnsEMBL::Variation::Utils::BaseVepPlugin);
+
+our $CAN_USE_FAIDX;
+our $DEBUG = 0;
+
+BEGIN {
+  if (eval { require Bio::DB::HTS::Faidx; 1 }) {
+    $CAN_USE_FAIDX = 1;
+  }
+}
+
+sub new {
+  my $class = shift;
+  
+  my $self = $class->SUPER::new(@_);
+
+  # get file
+  my $fasta_file = $self->params->[0] || "";
+  die("ERROR: Ancestral FASTA file $fasta_file not provided or found\n") unless $fasta_file && -e $fasta_file;
+  $self->fasta_file($fasta_file);
+
+  # get type of index we're creating/using
+  my $type = $fasta_file =~ /\.gz$/ ? 'Bio::DB::HTS::Faidx' : 'Bio::DB::Fasta';
+  die("ERROR: Bio::DB::HTS required to access compressed FASTA file $fasta_file\n") if $type eq 'Bio::DB::HTS::Faidx' && !$CAN_USE_FAIDX;
+  $self->index_type($type);
+
+  # init DB here so indexing doesn't happen at some point in the middle of VEP run
+  $self->fasta_db;
+
+  return $self;
+}
+
+sub feature_types {
+  return ['Feature','Intergenic'];
+}
+
+sub get_header_info {
+  return { AA => 'Ancestral allele from '.$_[0]->fasta_file }
+}
+
+sub run {
+  my ($self, $tva) = @_;
+  
+  my $vf = $tva->variation_feature;
+
+  my ($start, $end) = ($vf->{start}, $vf->{end});
+
+  # cache AA on VF, prevents repeat lookups for multiple TVAs
+  if(!exists($vf->{_vep_plugin_AncestralAllele})) {
+
+    # insertion
+    return { AA => $vf->{_vep_plugin_AncestralAllele} = '-' } if $start > $end;
+
+    my $sr_name = $self->get_sr_name($vf->{chr});
+
+    # chr not in FASTA
+    return { AA => $vf->{_vep_plugin_AncestralAllele} = '?' } unless $sr_name;
+
+    my $fasta_db = $self->fasta_db;
+    my $type = $self->index_type;
+    my ($seq, $length);
+
+    if($type eq 'Bio::DB::HTS::Faidx') {
+      my $location_string = $sr_name.":".$start."-".$end ;
+      ($seq, $length) = $fasta_db->get_sequence($location_string);
+    }
+    elsif($type eq 'Bio::DB::Fasta') {
+      $seq = $fasta_db->seq($sr_name, $start => $end);
+    }
+    else {
+      throw("ERROR: Don't know how to fetch sequence from a ".ref($fasta_db)."\n");
+    }
+
+    reverse_comp(\$seq) if $vf->{strand} < 0;
+
+    $vf->{_vep_plugin_AncestralAllele} = $seq;
+  }
+
+  return { AA => $vf->{_vep_plugin_AncestralAllele} };
+}
+
+# getter/setter for fasta file path
+sub fasta_file {
+  my $self = shift;
+  $self->{_fasta_file} = shift if @_;
+  return $self->{_fasta_file};
+}
+
+# getter/setter for fasta index type
+sub index_type {
+  my $self = shift;
+  $self->{_index_type} = shift if @_;
+  return $self->{_index_type};
+}
+
+## CRIBBED FROM Bio::EnsEMBL::Variation::Utils::FastaSequence
```

Comment:
I suggest this  the code based on existing module is moved to a shared utils space.